src/asm3/stock.py

```diff
@@ -122,6 +324,186 @@ def update_stocklevel_from_form(dbo: Database, post: PostedData, username: str)
+    """
+    Inserts a product item from a dialog.
+    """
+    l = dbo.locale
+    if post["productname"] == "":
+        raise asm3.utils.ASMValidationError(_("Product must have a name", l))
+    
+    retired = 0
+    if post.integer("active") == 0:
+        retired = 1
+   
+    pid = dbo.insert("product", {
+        "ProductName":          post["productname"],
+        "Description":          post["productdescription"],
+        "ProductTypeID":        post.integer("producttypeid"),
+        "SupplierID":           post.integer("supplierid"),
+        "UnitTypeID":           post.integer("unittypeid"),
+        "CustomUnit":           post["customunit"],
+        "PurchaseUnitTypeID":   post.integer("purchaseunittypeid"),
+        "CustomPurchaseUnit":   post["custompurchaseunit"],
+        "CostPrice":            post.integer("costprice"),
+        "RetailPrice":          post.integer("retailprice"),
+        "UnitRatio":            post.integer("unitratio"),
+        "TaxRateID":            post.integer("taxrateid"),
+        "IsRetired":            retired,
+        "Barcode":              post["barcode"],
+        "PLU":                  post["plu"],
+        "GlobalMinimum":        post["globalminimum"],
+        "RecentBatchNo":        "",
+        "RecentExpiry":         ""
+    }, username)
+
+    return pid
+
+def insert_productmovement_from_form(dbo: Database, post: PostedData, username: str) -> int:
+    """
+    Inserts a product movement from a dialog.
+    """
+    movementusagetypeid = asm3.configuration.product_movement_usage_type(dbo)
+    l = dbo.locale
+    if post["movementdate"] == "":
+        raise asm3.utils.ASMValidationError(_("Movement must have a date", l))
+    
+    locations = []
+
+    if post.integer("movementfromtype") == 0 and post.integer("movementtotype") == 1:# Stock to usage
+        locations = [post.integer("movementfrom"), False]
+        usagetypeid = post.integer("movementto")
+    elif post.integer("movementfromtype") == 0 and post.integer("movementtotype") == 0:# Stock to stock
+        locations = [post.integer("movementfrom"), post.integer("movementto")]
+        usagetypeid = movementusagetypeid
+    elif post.integer("movementfromtype") == 1 and post.integer("movementtotype") == 0:# Usage to stock
+        locations = [False, post.integer("movementto"),]
+        usagetypeid = post.integer("movementfrom")
+    else:# Usage to usage
+        locations = [post.integer("movementfrom"), post.integer("movementto")]
+        usagetypeid = movementusagetypeid
+    
+    quantity = post.integer("movementquantity")
+    unitratio = post.integer("unitratio")
+
+    # Get current stock levels of the selected product
+    stocklevels = dbo.query("SELECT ID, BatchNumber, Balance, Cost, UnitPrice, Total, Low, Expiry, UnitPrice, StockLocationID FROM stocklevel WHERE ProductID = %s ORDER BY Balance" % (post.integer("productid"),))
```

Comment:
Break this query up onto multiple lines. Always use parameterised queries where possible for security - the end of this line should be ```ProductID = ? ORDER BY Balance", [ post.integer("productid") ])```